cmd/kubeless/function.go

```diff
@@ -100,14 +119,246 @@ func parseMemory(mem string) (resource.Quantity, error) {
+		if err != nil {
+			return strconv.Itoa(i), nil
+		}
+		conn.Close()
+	}
+	return "", errors.New("Can not find an unassigned port")
+}
+
+const (
+	maxRetries       = 5
+	defaultTimeSleep = 1
+)
+
+func waitForHTTPServer(host, port string) error {
+	// Build the request
+	req, err := http.NewRequest(
+		"GET",
+		"http://"+host+":"+port,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Execute the request
+	resp, err := http.DefaultClient.Do(req)
+	retries := 0
+	for err != nil && retries < maxRetries {
+		retries++
+		time.Sleep(time.Duration(time.Second))
+		resp, err = http.DefaultClient.Do(req)
+	}
+	resp.Body.Close()
+	return err
+}
+
+func runPortForward(f cmdutil.Factory, k8sClient *rest.RESTClient, podName, port string) (err error) {
+	clientset, err := f.ClientSet()
+	if err != nil {
+		return err
+	}
+	k8sClientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	portSlice := []string{port + ":9000"}
+	go func() {
+		devNull, err := os.Open(os.DevNull)
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		pfo := k8scmd.PortForwardOptions{
+			RESTClient: k8sClient,
+			Namespace:  "kubeless",
+			Config:     k8sClientConfig,
+			PodName:    podName,
+			PodClient:  clientset.Core(),
+			Ports:      portSlice,
+			PortForwarder: &defaultPortForwarder{
+				cmdOut: devNull,
+				cmdErr: os.Stderr,
+			},
+			StopChannel:  make(chan struct{}, 1),
+			ReadyChannel: make(chan struct{}),
+		}
+		err = pfo.RunPortForward()
+		if err != nil {
+			logrus.Fatalf("Connection failed: %v", err)
+		}
+
+	}()
+	return
+}
+
+func getMinioURLFromOutsideCluster() (url string, err error) {
+	f := cmdutil.NewFactory(nil)
+	k8sClient, err := f.RESTClient()
+	if err != nil {
+		return
+	}
+
+	pods := &v1.PodList{}
+	pod := v1.Pod{}
+	k8sClient.Get().Namespace("kubeless").Resource("pods").Do().Into(pods)
+	for i := range pods.Items {
+		if pods.Items[i].Labels["kubeless"] == "minio" {
+			pod = pods.Items[i]
+			break
+		}
+	}
+
+	port, err := getLocalPort()
+	if err != nil {
+		return
+	}
+
+	// Minio requires to serve content at "root" eg. minio.com:9000
+	// so instead of using the Kubernetes Proxy we need to run
+	// a port forward to Minio's pod
+	runPortForward(f, k8sClient, pod.Name, port)
+	err = waitForHTTPServer("localhost", port)
+	if err != nil {
+		return
+	}
+	url = "localhost:" + port
+	return
+}
+
+func getMinioObject(minioClient minio.Client, bucket, name string) (object []byte, err error) {
+	var dir string
+	dir, err = ioutil.TempDir("", "")
+	if err != nil {
+		return
+	}
+	defer os.RemoveAll(dir)
+
+	err = minioClient.FGetObject(bucket, name, path.Join(dir, name), minio.GetObjectOptions{})
+	if err != nil {
+		return
+	}
+	object, err = ioutil.ReadFile(path.Join(dir, name))
+	return
+}
+
+func getMinioClient(endpoint string) (minioClient *minio.Client, err error) {
+	cli, err := utils.GetRestClientOutOfCluster("", "v1", "/api")
+	if err != nil {
+		return
+	}
+	result := &v1.Secret{}
+	cli.Get().
```

Comment:
Prefer the (typed) client in `k8s.io/client-go/kubernetes` for standard resources over the low-level `rest.Client`, since those wrappers will allow golang to catch additional classes of errors at compile time.